### PR TITLE
Adds FLUX attention masking and additional time embedder

### PR DESCRIPTION
**Pull request by clementchadebec**
## Added features
- Adds attention masking to Flux
- Add the possibility to add an extra vector embedding added to the timesteps embedding for Unet and FluxTransformer

### CODE REVIEW

File: src/diffusers/models/transformers/transformer_flux.py [hunk 1]
```diff
@@ -20,7 +20,11 @@
 import torch.nn as nn
 
 from ...configuration_utils import ConfigMixin, register_to_config
-from ...loaders import FluxTransformer2DLoadersMixin, FromOriginalModelMixin, PeftAdapterMixin
+from ...loaders import (
```
Comment (benjaminaubin) [inline src/diffusers/models/transformers/transformer_flux.py:23]: Did something change in the black formatting?

File: src/diffusers/models/transformers/transformer_flux.py [hunk 2]
```diff
@@ -455,6 +534,14 @@ def forward(
             if guidance is None
             else self.time_text_embed(timestep, guidance, pooled_projections)
         )
+
+        if discretization is not None:
+            discretization = self.disc_proj(
+                discretization
+            )  # * (discretization > 0).unsqueeze(1)
```
Comment (benjaminaubin) [inline src/diffusers/models/transformers/transformer_flux.py]: remove comment ?

File: src/diffusers/models/transformers/transformer_flux.py [hunk 3]
```diff
@@ -455,6 +534,14 @@ def forward(
             if guidance is None
             else self.time_text_embed(timestep, guidance, pooled_projections)
         )
+
+        if discretization is not None:
```
Comment (benjaminaubin) [inline src/diffusers/models/transformers/transformer_flux.py]: Change name to `additional_timestep_embd` or something like this

File: src/diffusers/models/unets/unet_2d_condition.py [hunk 4]
```diff
@@ -235,6 +235,7 @@ def __init__(
         mid_block_only_cross_attention: Optional[bool] = None,
         cross_attention_norm: Optional[str] = None,
         addition_embed_type_num_heads: int = 64,
+        additional_class_embeddings: bool = False,
```
Comment (benjaminaubin) [inline src/diffusers/models/unets/unet_2d_condition.py]: I would unify the notation between unet and the transformer

File: src/diffusers/models/transformers/transformer_flux.py [hunk 5]
```diff
@@ -250,12 +301,23 @@ def __init__(
...
+            else CombinedTimestepTextProjEmbeddings
         )
+
         self.time_text_embed = text_time_guidance_cls(
             embedding_dim=self.inner_dim, pooled_projection_dim=pooled_projection_dim
         )
 
+        if discretization_embeds:
+            self.disc_proj = Timesteps(
+                num_channels=256, flip_sin_to_cos=True, downscale_freq_shift=0
```
Comment (benjaminaubin) [inline src/diffusers/models/transformers/transformer_flux.py:315]: shall those parameters be passed in the config? (this is not harcoded in the unet for instance)

File: src/diffusers/models/unets/unet_2d_condition.py [hunk 6]
```diff
@@ -1296,6 +1309,16 @@ def forward(
...
 
+        if additional_class_labels is not None:
+            additional_class_emb_proj = self.time_proj(additional_class_labels).to(
+                dtype=emb.dtype
+            )
+            additional_class_emb = self.additional_class_embedding(
+                additional_class_emb_proj
+            )
+            emb = emb + additional_class_emb
+            print("additional_class_emb", emb.shape)
```
Comment (benjaminaubin) [inline src/diffusers/models/unets/unet_2d_condition.py]: Replace by logging.debug

**Review by benjaminaubin — CHANGES_REQUESTED** [review]
Hi @clementchadebec , thank you for sharing those modifications to allow anyone playing with Kontext

File: src/diffusers/models/unets/unet_2d_condition.py (hunk 6 above)
Comment (clementchadebec) [reply src/diffusers/models/unets/unet_2d_condition.py]: removed

File: src/diffusers/models/transformers/transformer_flux.py (hunk 5 above)
Comment (clementchadebec) [reply src/diffusers/models/transformers/transformer_flux.py:315]: This is also hard-coded for the original timestep embedder 
https://github.com/initml/diffusers/blob/e82eefd651aaa73243ae4ce808b47ef5310660fb/src/diffusers/models/embeddings.py#L1620-L1623

**Comment by clementchadebec** [comment]
I removed the changes `unet_2d_condition.py` since they were not useful and not tested for now

**Review by benjaminaubin — APPROVED** [review]
